File: src/js/packages/common/src/variants/pipeline.ts
import { ArtifactSource } from "../artifact.js";
import { DtypesManager } from "../dtypes.js";
import { InvalidArtifactFileError, OperationInstanceError } from "../errors.js";
import { Manifest, OpInstanceConfig, OpIO, PipelineVariant } from "../interfaces.js";
import { BaseOp } from "../ops/base.js";
import Registry from "../registry.js";
import { BaseVariant } from "./base.js";
import { DagComponent, dagCompute } from "./common/dag.js";
import { validateInputs } from "./common/validators.js";

export function validatePipeline(
    manifest: Manifest,
    ops: OpInstanceConfig[],
    variantConfig: Record<string, unknown>
): void {
    for (const [entryKind, entries] of [
        ["input", manifest.inputs],
        ["output", manifest.outputs],
    ] as const) {
        for (const entry of entries) {
            if (entry.content_type !== "NDJSON") {
                throw new InvalidArtifactFileError(
                    "manifest.json",
                    `Pipeline ${entryKind} \`${entry.name}\` requires the NDJSON content type, ` +
                        `got \`${entry.content_type}\``
                );
            }
        }
    }

    if (!Array.isArray(variantConfig.nodes)) {
        throw new InvalidArtifactFileError("variant_config.json", "`nodes` must be an array");
    }

    const opInstances = new Map(ops.map((instance) => [instance.id, instance]));
    const boundNames = new Set<string>();
    for (const input of manifest.inputs) {
        if (boundNames.has(input.name)) {
            throw new Error(`Pipeline input \`${input.name}\` binds a value more than once`);
        }
        boundNames.add(input.name);
    }

    const config = variantConfig as unknown as PipelineVariant;
    for (const [nodeIndex, node] of config.nodes.entries()) {
        const nodeName = `Pipeline node ${nodeIndex} (\`${node.op_instance_id}\`)`;
        const opInstance = opInstances.get(node.op_instance_id);
        if (!opInstance) {
            throw new Error(
                `${nodeName} references undeclared op instance \`${node.op_instance_id}\``
            );
        }

        for (const ioKind of ["inputs", "outputs"] as const) {
            const nodeArity = node[ioKind].length;
            const opArity = opInstance[ioKind].length;
            if (nodeArity !== opArity) {
                throw new Error(
                    `${nodeName} has ${ioKind.slice(0, -1)} arity ${nodeArity}, but op ` +
                        `instance \`${node.op_instance_id}\` declares ${opArity}`
                );
            }
        }

        for (const inputName of node.inputs) {
            if (!boundNames.has(inputName)) {
                throw new Error(`${nodeName} consumes unbound input \`${inputName}\``);
            }
        }
        for (const outputName of node.outputs) {
            if (boundNames.has(outputName)) {
                throw new Error(`${nodeName} binds value \`${outputName}\` more than once`);
            }
            boundNames.add(outputName);
        }
    }
}

class PipelineNodeInstance implements DagComponent {
    readonly operator: BaseOp;
    readonly inputs: string[];
    readonly outputs: string[];
    readonly inputSpecs: OpIO[];
    readonly outputSpecs: OpIO[];
    readonly extra_dynattrs: Record<string, string>;

    constructor(config: {
        operator: BaseOp;
        inputs: string[];
        outputs: string[];
        inputSpecs: OpIO[];
        outputSpecs: OpIO[];
        extra_dynattrs: Record<string, string>;
    }) {
        this.operator = config.operator;
        this.inputs = config.inputs;
        this.outputs = config.outputs;
        this.inputSpecs = config.inputSpecs;
        this.outputSpecs = config.outputSpecs;
        this.extra_dynattrs = config.extra_dynattrs;
    }
}

export class Pipeline extends BaseVariant {
    private readonly pipelineNodeInstances: PipelineNodeInstance[];

    constructor(
        source: ArtifactSource,
        ops: OpInstanceConfig[],
        variantConfig: Record<string, unknown>,
        registry: Registry,
        dtypesManager: DtypesManager
    ) {
        super(source, ops, variantConfig, registry, dtypesManager);
        const config = variantConfig as unknown as PipelineVariant;
        this.pipelineNodeInstances = config.nodes.map((node) => {
            const opInstance = this.opInstances.get(node.op_instance_id);
            if (!opInstance) {
                throw new OperationInstanceError(node.op_instance_id, "instance is not declared");
            }
            return new PipelineNodeInstance({
                operator: opInstance.operator,
                inputs: node.inputs,
                outputs: node.outputs,
                inputSpecs: opInstance.inputSpecs,
                outputSpecs: opInstance.outputSpecs,
                extra_dynattrs: node.extra_dynattrs ?? {},
            });
        });
    }

    private async nodeCompute(
        nodeInstance: PipelineNodeInstance,
        nodeInputs: unknown[],
        passthrough: Record<string, unknown>
    ): Promise<unknown> {
        validateInputs(nodeInputs, nodeInstance.inputSpecs);
        return nodeInstance.operator.compute(
            nodeInputs,
            passthrough.dynamic_attributes as Record<string, string>
        );
    }

    async compute(
        inputs: Record<string, unknown>,
        dynamicAttributes: Record<string, string>
    ): Promise<Record<string, unknown>> {
        return dagCompute(
            inputs,
            this.pipelineNodeInstances,
            (component, values, passthrough) =>
                this.nodeCompute(
                    component as PipelineNodeInstance,
                    values,
                    passthrough as Record<string, unknown>
                ),
            (result) => (result as { value: unknown[] }).value,
            { dynamic_attributes: dynamicAttributes }
        );
    }
}

File: src/python/fnnx/variants/pipeline.py
from dataclasses import dataclass
from typing import Any

from fnnx.node_instance import OpInstance
from fnnx.variants._base import BaseVariant
from fnnx.variants._common.dag import DagComponent, dag_compute, dag_compute_async
from fnnx.variants._common.validators import validate_inputs


@dataclass
class PipelineNodeInstance(OpInstance, DagComponent):
    pass


def validate_pipeline(
    manifest: dict[str, Any],
    op_instances: list[dict[str, Any]],
    variant_config: dict[str, Any],
) -> None:
    for entry_kind in ("input", "output"):
        for entry in manifest[f"{entry_kind}s"]:
            if entry["content_type"] == "JSON":
                raise ValueError(
                    f"Pipeline {entry_kind} `{entry['name']}` cannot use JSON content type"
                )

    instances_by_id = {instance["id"]: instance for instance in op_instances}
    bound_names: set[str] = set()
    for input_entry in manifest["inputs"]:
        input_name = input_entry["name"]
        if input_name in bound_names:
            raise ValueError(
                f"Pipeline input `{input_name}` binds a value more than once"
            )
        bound_names.add(input_name)

    for node_index, node in enumerate(variant_config["nodes"]):
        op_instance_id = node["op_instance_id"]
        node_name = f"Pipeline node {node_index} (`{op_instance_id}`)"
        op_instance = instances_by_id.get(op_instance_id)
        if op_instance is None:
            raise ValueError(
                f"{node_name} references undeclared op instance `{op_instance_id}`"
            )

        for io_kind in ("inputs", "outputs"):
            node_arity = len(node[io_kind])
            op_arity = len(op_instance[io_kind])
            if node_arity != op_arity:
                raise ValueError(
                    f"{node_name} has {io_kind[:-1]} arity {node_arity}, but op "
                    f"instance `{op_instance_id}` declares {op_arity}"
                )

        for input_name in node["inputs"]:
            if input_name not in bound_names:
                raise ValueError(f"{node_name} consumes unbound input `{input_name}`")

        for output_name in node["outputs"]:
            if output_name in bound_names:
                raise ValueError(
                    f"{node_name} binds value `{output_name}` more than once"
                )
            bound_names.add(output_name)


class Pipeline(BaseVariant):
    def _post_init(
        self,
    ) -> None:
        self.pipeline_node_instances: list[PipelineNodeInstance] = []
        for node in self.variant_config["nodes"]:
            op_instance = self.op_instances[node["op_instance_id"]]
            self.pipeline_node_instances.append(
                PipelineNodeInstance(
                    operator=op_instance.operator,
                    inputs=node["inputs"],
                    outputs=node["outputs"],
                    input_specs=op_instance.input_specs,
                    output_specs=op_instance.output_specs,
                    extra_dynattrs=node.get("extra_dynattrs", {}),
                )
            )

    async def _node_compute_async(
        self,
        node_instance: PipelineNodeInstance,
        node_inputs: list[Any],
        **node_passtrhough: Any,
    ) -> Any:
        validate_inputs(node_inputs, node_instance.input_specs)
        return await node_instance.operator.compute_async(
            node_inputs, **node_passtrhough
        )

    def _node_compute(
        self,
        node_instance: PipelineNodeInstance,
        node_inputs: list[Any],
        **node_passtrhough: Any,
    ) -> Any:
        validate_inputs(node_inputs, node_instance.input_specs)
        return node_instance.operator.compute(node_inputs, **node_passtrhough)

    async def compute_async(
        self,
        inputs: dict[str, Any],
        dynamic_attributes: dict[str, str],
    ) -> dict[str, Any]:
        passthrough = {
            "op_executor": self.op_executor,
            "dynamic_attributes": dynamic_attributes,
        }

        return await dag_compute_async(
            inputs,
            self.pipeline_node_instances,
            self._node_compute_async,
            as_val=lambda res: res.value,
            components_passthrough=passthrough,
        )

    def compute(
        self,
        inputs: dict[str, Any],
        dynamic_attributes: dict[str, str],
    ) -> dict[str, Any]:
        passthrough = {
            "op_executor": self.op_executor,
            "dynamic_attributes": dynamic_attributes,
        }

        return dag_compute(
            inputs,
            self.pipeline_node_instances,
            self.executor,
            self._node_compute,
            as_val=lambda res: res.value,
            components_passthrough=passthrough,
        )
